Basic implementation of new NMR schema

diptobi1 opened the pull request:
Added excusedMissedTurns as option on game creation
Added display of excused turns and used up turns per member in game panel
Added excusedMissedTurns as search item in game search

diptobi1 committed:
Adjusted SQL update to include default values for new columns in wD_Games, wD_Users, wD_Members and increased webdip version according to DB value

diptobi1 committed:
Add excusedMissedTurns as option in game creation

diptobi1 committed:
Display excused NMRs and used up excuses in gamepanel

diptobi1 committed:
Added Excused missing turs in game search

diptobi1 committed:
Fix typos in gamemaster/game.php

diptobi1 committed:
Improved display of NMR excuses in gamepanel

diptobi1 commented:
I focused here on game creation and display in game panel as a place to start. Just a few lines. Let me know if there are any problems with the commits.

diptobi1 committed:
Restructured NMR handling in process game

jmo1121109 commented:
At first look through this looks great. I've started going through in detail but haven't finished since you included a lot! Thanks for this!

diptobi1 commented:
Just a note on the last commit:

I split the tasks for NMR handling in a few functions so the normal process routine only gets executed if no NMRs by playing members appeared and certain NMR handling only takes place other wise. Also tasks not directly related to the record of NMRs got there own method to make the code clearer.

Note that the last commit is probably not runnable yet and not even partly tested. Just wanted to push it this evening as I will not have time to finish this for the next few days.

jmo1121109 commented on gamepanel/member.php:
```diff
@@ -442,18 +450,24 @@ function memberUserDetail()
+				if($this->Game->excusedMissedTurns > 0)
+					$buf .= ' - '.l_t('NMRs: %s of %s','<span class="excusedNMRs">'.$this->excusedMissedTurns.'</span>','<span class="excusedNMRs">'.$this->Game->excusedMissedTurns.'</span>');
+				
+				if ( $this->missedPhases == 2 )
```
I'm not sure I'm understanding what this section of the code is going to be doing. We should never get to a counter of missedPhases == 2 right? At that point they should have been kicked out of the game if they were out of excused missed turns?

diptobi1 replied:
The missedPhases part is actually old code that I just left there for the time being to display missed phases as before. At first I was confused about the 2, as well, though the old code works in the following way: After a turn is processed all members missed phases counter is increased by 1 and is only reset to 0 if the member loads the board. So two missed phases in the counter actually mean, that the member was absent during the whole turn before.

Though, as we could have multiple misses in a row, this should be replaced in any case. I guess it would be easiest to keep this missed Phases counter and just replace the "==" by ">=" and increase the counter in any case at the end of the process method. Working instead with the actual database entries could be a headache here as the turn does not progress and we would have to work with the turnAsDate and all the complications of extended phases etc..

jmo1121109 commented on install/1.44-1.45/update.sql:
```diff
@@ -16,11 +16,11 @@ CREATE TABLE IF NOT EXISTS `wD_MissedTurns` (
-ALTER TABLE `wD_Users` ADD `yearlyMissedTurnCount` mediumint(8) unsigned;
-ALTER TABLE `wD_Users` ADD `shortTermMissedTurnCount` mediumint(8) unsigned;
-ALTER TABLE `wD_Users` ADD `yearlyPhaseCount` mediumint(8) unsigned;     
+ALTER TABLE `wD_Members` ADD `excusedMissedTurns` int(10) unsigned DEFAULT 0;
```
Oh this is interesting, I had not considered that this is going to impact ongoing games when we do this change. Do you think we should default existing games to 1 or 2 excuses to "simulate" being able to NMR twice currently before being forced into CD like the sites currently do?

diptobi1 replied:
I have chosen 2 as default in game creation and would go with 2 here, as well. With 2 excuses you will pushed into CD after the third NMR, which seems to be a good middle way between the current situation of CD after 2 consecutive NMRs and infinite not consecutive ones.

jmo1121109 commented on objects/member.php:
```diff
@@ -103,6 +103,11 @@ class Member
 	var $missedPhases;
+	
+	/**
+	 * The number of already used up NMRs by this Member that got excused
```
Based on how this is used above "SET m.excusedMissedTurns = m.excusedMissedTurns - 1" I think this will need to hold 

/**
* The number of excused turns this member has left.

diptobi1 replied:
Yes, I already realized that I assumed a different definition of member's excused missed turns than you (counting up not down). I will need to adjust the initial value on game creation and probably the display on the board page, as well.

jmo1121109 reviewed (changes requested):
This looks really really good so far. Thanks for making this commit! I had a couple very minor things and a question about how we should handle something and then I'll merge this in.

diptobi1 commented:
Thanks for your review. I will not have to time to put hands on this for probably another week. So if you would like to progress with my changes before, you might need to adjust the commits yourself. Otherwise I will change it when there is time again.

jmo1121109 commented:
> Thanks for your review. I will not have to time to put hands on this for probably another week. So if you would like to progress with my changes before, you might need to adjust the commits yourself. Otherwise I will change it when there is time again.

Oh no rush at all with this, it's a larger project and I'm happy to take our time on it to get it working smoothly. There's plenty of other small changes I'm trying to clean up around the sites too in-between.

diptobi1 committed:
- Altered code to count member excuses down not up
- CD takeovers get excuses (TODO: Handling of takeovers from members who left before)
- Adjusted old NMR code to count missed phases even before adjudication (if an NMR happened, the game will not be progressed so there is no need to do it after the processing in case of a crash) and to reset it if no NMR happened; removed code to reset the count when the member returns to the game (the mere purpose of this count is now to notify other  members about the NMR)
- Small fixes to get the last commit actually running

diptobi1 committed:
Basic implementation of the new NMR schema:

Members with no complete set of orders lose an excuse.

If no excuses are left, members go into CD.

The phase only progesses if there are orders from each country, that has to order.

Members are notified about phase extensions and use up of excuses.

TODO: Further handling of CDs that goes beyond the current state (temp ban etc).
Code can be directly added to Member->setLeft

diptobi1 commented:
Your concerns are now addressed: 

Excuses are now counted down and displayed as such everywhere. A default value of two excuses for running games is now implemented.

I decided to keep the missedPhases as indicator for a missed deadline. Though I altered the code a bit to make it more understandable as the old main purpose of this value to detect to consecutive NMRs is not needed anymore. The missed turns are now registered together with all other NMR stuff and not at the end of the processing. As there is no processing when NMRs happen, we do not have to worry anymore, if the processing crashes that could have lead to multiple NMR recordings before.

Furthermore I continued to work on the NMR handling. We are now at a state, where NMRs are tracked correctly and members are pushed into CD when excuses are used up (no temp ban stuff yet). Members are also notified about extensions etc.. I decided to unready all orders when a phase is extended so each player has a chance to recheck the game before a position is overtaken and readied up. But we could easily change that if there is a reason to do so.

jmo1121109 commented:
This looks good so far, still going through it. I'm not super familiar with the game engine so trying to make sure I really understand it.

diptobi1 commented:
Alright. Already realized that I was a bit sloppy so far concerning players who are already left. The next commit fixing this and implementing automatic temp bans is already in the pipeline. Just need to do some debugging.

jmo1121109 reviewed (approved):
Okay finally got through everything! This looks fantastic, the code refactoring is really good too. Merging this into my branch.

diptobi1 commented:
Thanks for the review. 
I already have temp ban added as the next commit and am currently working on preventing rejoins during a temp ban (so a temp banned members cannot postpone their games infinitely). Will wait with my next push until I finished that work.

jmo1121109 commented:
hahaha you're writing this almost faster then I can review it. Awesome job. Between this and my styling work we might have webdip and vdip looking and behaving like modern sites soon.
